File: Assets/script/muzzle.js
#pragma strict
var attackObjet : GameObject;
var trigger : String;

function Start () {
    trigger = transform.parent.gameObject.name;
}

function Update () {
    //get parentName And Fire
    if(Input.GetButtonDown(trigger)){
         var attack : GameObject = Instantiate(attackObjet,transform.position,transform.rotation);
         attack.transform.parent = this.transform;
         attack.BroadcastMessage("setSpeed",SendMessageOptions.DontRequireReceiver);
    }
    var mouse : Vector3 = new Vector3(Input.mousePosition.x,0,Input.mousePosition.z);

}

File: Assets/script/Unit/muzzle.cs
using UnityEngine;
using System.Collections;

public class muzzle : MonoBehaviour {

    public GameObject attackObject;
    private string trigger;

	// Use this for initialization
	void Start () {
        
	}
	
	// Update is called once per frame
	void Update () {
        if (transform.parent.tag == "Player" && Input.GetButtonDown("Fire1"))
        {
            transform.parent.transform.parent.gameObject.BroadcastMessage("fireCheck",gameObject);
        }
        else
        {
            //TODO enemy attack
        }
	}

    public void fire()
    {
        GameObject attack = (GameObject)Instantiate(attackObject, transform.position, transform.rotation);
        attack.transform.parent = this.transform;
        attack.BroadcastMessage("StartAttack", SendMessageOptions.DontRequireReceiver);
    }
}
